File: server/app/models/category.py
from sqlalchemy import Column, Integer, String, TIMESTAMP, func, ForeignKey
from sqlalchemy.orm import relationship

from app.database import Base


class Category(Base):
    __tablename__ = "categories"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, nullable=False)
    slug = Column(String, index=True)

    level = Column(Integer, default=0)
    parent_id = Column(Integer, ForeignKey("categories.id"), nullable=True)

    children = relationship("Category", foreign_keys=[parent_id], lazy="joined")

    created_at = Column(TIMESTAMP(timezone=True),
                        nullable=False, server_default=func.now())
    updated_at = Column(TIMESTAMP(timezone=True),
                        default=None, onupdate=func.now())

File: client/types/category.ts
import type { BlockRsp } from '~/types/base';

export type CategoryData = {
    level: number;
    id: number;
    created_at: Date;
    slug: string;
    name: string;
    parent_id: number | null;
    updated_at: Date | null;
    children?: CategoryData[];
};

export type CategoriesRsp = BlockRsp<CategoryData[]>;
